src/jdk.jdwp.agent/share/native/libjdwp/util.c

```diff
@@ -1099,31 +1106,425 @@ debugMonitorNotifyAll(jrawMonitorID monitor)
+    jvmtiError error;
+    error = JVMTI_FUNC_PTR(gdata->jvmti,RawMonitorExit)
+            (gdata->jvmti, dbgRawMonitor);
+    error = ignore_vm_death(error);
+    if (error != JVMTI_ERROR_NONE) {
+      EXIT_ERROR(error, "on dbgRawMonitor exit");
+    }
+}
+
+DebugRawMonitor*
+debugMonitorCreate(DebugRawMonitorRank dbg_monitor_rank, char *name)
 {
     jrawMonitorID monitor;
     jvmtiError error;
+    DebugRawMonitor* dbg_monitor = &dbg_monitors[dbg_monitor_rank];
+    JDI_ASSERT(dbg_monitor->monitor == NULL);
 
     error = JVMTI_FUNC_PTR(gdata->jvmti,CreateRawMonitor)
                 (gdata->jvmti, name, &monitor);
     if (error != JVMTI_ERROR_NONE) {
         EXIT_ERROR(error, "on creation of a raw monitor");
     }
-    return monitor;
+
+    JDI_ASSERT(monitor != NULL);
+
+    // Need to lock during initialization so verifyMonitorRank() can be guaranteed that
+    // if the monitor field is not NULL, then the monitor is fully initialized.
+    if (gdata->rankedMonitors) {
+        dbgRawMonitor_lock();
+    }
+    dbg_monitor->monitor = monitor;
+    dbg_monitor->rank = dbg_monitor_rank;
+    dbg_monitor->name = name;
+    if (gdata->rankedMonitors) {
+        dbgRawMonitor_unlock();
+    }
+
+    return dbg_monitor;
 }
 
 void
-debugMonitorDestroy(jrawMonitorID monitor)
+debugMonitorDestroy(DebugRawMonitor *dbg_monitor)
 {
     jvmtiError error;
+    JDI_ASSERT(dbg_monitor != NULL);
 
     error = JVMTI_FUNC_PTR(gdata->jvmti,DestroyRawMonitor)
-                (gdata->jvmti, monitor);
+                (gdata->jvmti, dbg_monitor->monitor);
     error = ignore_vm_death(error);
     if (error != JVMTI_ERROR_NONE) {
         EXIT_ERROR(error, "on destruction of raw monitor");
     }
+
+    dbg_monitor->monitor = NULL;
+}
+
+static char*
+getThreadName(jthread thread)
+{
+    jvmtiThreadInfo info;
+    jvmtiError error;
+    JNIEnv *env = getEnv();
+
+    memset(&info, 0, sizeof(info));
+    WITH_LOCAL_REFS(env, 2) {
+        error = JVMTI_FUNC_PTR(gdata->jvmti,GetThreadInfo)
+                (gdata->jvmti, thread, &info);
+    } END_WITH_LOCAL_REFS(env);
+    return info.name;
+}
+
+void
+dumpRawMonitor(DebugRawMonitor *dbg_monitor) {
+    tty_message("DebugRawMonitor(%d)", dbg_monitor->rank);
+    tty_message("\tname: %s", dbg_monitor->name);
+    if (dbg_monitor->ownerThread != NULL) {
+        char* threadName = getThreadName(dbg_monitor->ownerThread);
+        tty_message("\townerThreadName: %s", threadName);
+        jvmtiDeallocate(threadName);
+    }
+}
+
+void
+dumpRawMonitors() {
+    int i;
+    for (i = 0; i < NUM_DEBUG_RAW_MONITORS; i++) {
+        DebugRawMonitor* dbg_monitor = &dbg_monitors[i];
+        if (dbg_monitor->monitor == NULL) {
+            continue;
+        }
+        jthread thread = dbg_monitor->ownerThread;
+        if (thread == NULL) {
+            continue;
+        }
+        dumpRawMonitor(dbg_monitor);
+    }
+}
+
+static void
+assertIsCurrentThread(JNIEnv *env, jthread thread, jthread current_thread)
```

Comment:
The function gets both threads as arguments, I think `assertIsSameThread` would be more correct name